Defered logging

suhasj opened the pull request:
@HaoK @divega 

Changes
- Fixed logging to pass a func instead of a string. This func is called in the logger only if the loglevel is enabled
- Fixed tests to pass func instead of string values as intended

divega commented on src/Microsoft.AspNet.Identity/RoleManager.cs:
```diff
@@ -404,12 +404,12 @@ private IRoleClaimStore<TRole> GetClaimStore()
             TRole role, [System.Runtime.CompilerServices.CallerMemberName] string methodName = "")
         {
-            result.Log(Logger, Resources.FormatLoggingResultMessageForRole(methodName, await GetRoleIdAsync(role)));
+            result.Log(Logger, () => { return Resources.FormatLoggingResultMessageForRole(methodName, GetRoleIdAsync(role).Result); });
```
You shouldn't need `return` here.

divega replied:
Hmmm. Sprinkling code that can potentially block everywhere isn't good. Let's chat about this more. We should probably hold of making this change until there is a version of `Log()` that takes an async lambda.

suhasj replied:
This is called in the IdentityResult.Log method https://github.com/aspnet/Identity/blob/DeferedLogging/src/Microsoft.AspNet.Identity/IdentityResult.cs#L65 so it has to return some result and since the log methods are not async I do a block 

These are called only if the loglevel is enabled though

divega replied:
Yo are missing my point. Try `() => Resources.FormatLoggingResultMessageForRole(...)`. Single line lambdas don't need `return` to return something.

suhasj replied:
ah i see. Thanks I got that. 

But the `.Result` call has to still be there

HaoK replied:
.Result is gross, we should just have it be LogAsync for now, and take a Func<Task<string>> so you can await it properly

suhasj commented:
@divega @HaoK 

Updated with async extension methods for logger and logging via fluent api. Some questions
- `WithLoggingAsync(Task<IdentityResult>)` calls `WithLoggingAsync(IdentityResult)` after calling an await in the later. can there be a way to beautify that.
- Should the method `WithLoggingAsync` be renamed to something else ? Some options I had
  * PostLoggingAsync
  * LogAndReturnAsync

suhasj commented:
@AndriySvyryd including you since @divega mentioned you could have thoughts on reducing excessive `async/await` in this case

HaoK commented on src/Microsoft.AspNet.Identity/SignInManager.cs:
```diff
@@ -64,13 +64,13 @@ public class SignInManager<TUser> where TUser : class
             if (Options.SignIn.RequireConfirmedPhoneNumber && !(await UserManager.IsPhoneNumberConfirmedAsync(user)))
             {
-                return await LogResultAsync(false, user);
+                return await false.WithLoggingAsync(this, user);
```
Personally I think this looks kind of ugly compared to the earlier iteration, extension off of false just seems really strange, also now we are passing this everywhere, which makes me think the methods really belong on the appropriate Manager

HaoK replied:
return await false.WithLoggingAsync just looks ugly to me, we are awaiting a bool?

suhasj replied:
I agree with that. We can a split where logging after Task<IdentityResult> can be an extension method while for bool results it can be in the Manager class itself.

Having an extension method makes it easy to have a fluent call that's all

HaoK replied:
There's a definite smell if you  are basically passing this to every log method.

divega replied:
Agreed that this one looks weird. Also it seems we could refactor the method to capture the outcome and three places and only log and return it at the end.

HaoK commented on src/Microsoft.AspNet.Identity/TaskExtensions.cs:
```diff
@@ -0,0 +1,44 @@
+        public static async Task<IdentityResult> WithLoggingAsync<TUser>(this Task<IdentityResult> identityResult, UserManager<TUser> userManager,TUser user,
+            [System.Runtime.CompilerServices.CallerMemberName] string methodName = "") where TUser :class
+        {
+            return await (await identityResult).WithLoggingAsync(userManager, user, methodName);
```
Ick await (await ...)  Yeah this definitely makes me think the extension method on the result types are the wrong way to go

suhasj replied:
sure that is one place I need feedback on. Definitely felt wrong adding that. Alternatively it would be `var result = await task` and then call the method on it which is the same thing.

One thing I looked into was `Task.ContinueWith` but the parameter was `Task<IdentityResult>` and not `IdentityResult` so the await had to be there anyway

HaoK replied:
My main feedback is that I think the switch to extension methods overall makes the code look  worse, and its not worth it, what you had before was better IMO, @divega what are your thoughts?

divega replied:
Well, await (await ...) just happens to be what you need sometimes although this is the one you could probably improve a bit with `ContinueWith()` (@suhasj note that the task in the delegate argument of `ContinueWith()` is already completed, or failed).

I don't completely hate the `WithLogging()` experiment. It has a good vibe in that it makes Logging a separate "aspect" that only applies to the results of operations. The one thing that makes me think it is isn't the right way to go is that on every single call we are passing the "this" argument, e.g. the UserManager, RoleManager, SignInManager instance. It is a clear smell that an instance method on those (the way we had it before) would make the most sense.

HaoK replied:
Yeah I said the same thing about the "this" smell. Also result extensions is no longer extensible unlike the virtual method. If they add their own derived sign in result with a new status code there is no way to update the logging extension method.

divega replied:
> Also result extensions is no longer extensible unlike the virtual method

Good point too.

divega commented on src/Microsoft.AspNet.Identity/LoggerExtensions.cs:
```diff
@@ -0,0 +1,74 @@
+        {
+            if (logger.IsEnabled(LogLevel.Information))
+            {
+                return logger.WriteAsync(LogLevel.Information, 0, state, exception, formatter);
```
This is invoking the `WriteAsync()` method defined above with a specific `LogLevel`, right? Why check if `logger.IsEnabled()` twice?

suhasj replied:
So this method is public and users can call it separately, so this additional layer of check. I can have public wrapper and make a private method without the check and remove the duplication

divega commented on src/Microsoft.AspNet.Identity/LoggerExtensions.cs:
```diff
@@ -0,0 +1,74 @@
+        {
+            if (logger.IsEnabled(LogLevel.Warning))
+            {
+                return logger.WriteAsync(LogLevel.Warning, 0, state, exception, formatter);
```
Same as above, why check for `logger.IsEnabled()` twice?

davidfowl commented:
I haven't really looked by why is there async logging?

HaoK commented:
There is no way to synchronously get the user name from identity. It's async only.  So the helper methods that generate log messages that include the user name have to be async. Logging itself doesn't need to be.

divega commented:
@davidfowl In the particular case of Identity, producing a helpful message to be sent to the log can involve async work, e.g. getting the username for a user is done invoking an async API (note that the default implementation isn't really async, but the design allows for it so se make no assumptions). This may apply or not to other frameworks.

divega commented on src/Microsoft.AspNet.Identity/SignInManager.cs:
```diff
@@ -214,7 +215,7 @@ private static ClaimsIdentity CreateIdentity(TwoFactorAuthenticationInfo info)
             var token = await UserManager.GenerateTwoFactorTokenAsync(user, provider);
             await UserManager.NotifyTwoFactorTokenAsync(user, provider, token);
-            return await LogResultAsync(true, user);
+            return await true.WithLoggingAsync(this, user);
```
Are we missing logging for the `return false` outcomes above?

suhasj replied:
The token is always generated so there is not false result as such hence omitted that

divega commented on src/Microsoft.AspNet.Identity/RoleManager.cs:
```diff
@@ -286,7 +286,7 @@ public virtual string NormalizeKey(string key)
             await Store.SetRoleNameAsync(role, name, CancellationToken);
             await UpdateNormalizedRoleNameAsync(role);
-            return await LogResultAsync(IdentityResult.Success, role);
+            return await Task.FromResult(IdentityResult.Success).WithLoggingAsync(this, role);
```
Why isn't this using the version of `WithLoggingAsync()` that takes an `IdentityResult` as opposed to `Task<IdentityResult>`?

divega replied:
Ah, because such version doesn't exist yet for RoleManager... I think it should exist.

divega replied:
Of course as a instance virtual method on `RoleManager`.

suhasj replied:
There is just one instance where we return IdentityResult instead of Task in RoleManager so I did not add that. I'll add that change

divega commented on src/Microsoft.AspNet.Identity/TaskExtensions.cs:
```diff
@@ -0,0 +1,44 @@
+
+namespace Microsoft.AspNet.Identity
+{
+    public static class TaskExtensions
```
I would put these alongside the ones in `ResultExtensions`. Results just happen to sometimes be async.

suhasj replied:
I thought we grouped extension methods based on the classes on which they hung on. I club all these log extension methods in one place ( if we end up doing them )

divega commented on src/Microsoft.AspNet.Identity/TaskExtensions.cs:
```diff
@@ -0,0 +1,44 @@
+            return await (await identityResult).WithLoggingAsync(userManager, user, methodName);
+        }
+
+        public static async Task<IdentityResult> WithLoggingAsync<TRole>(this Task<IdentityResult> identityResultDelegate, RoleManager<TRole> roleManager, TRole role,
```
Why the name `identityResultDelegate`? What makes it a delegate?

suhasj replied:
Because we were awaiting on it, I thought it was good to put the delegate in the name. Now I feel delegate makes sense only if it is a `Func` or an `Action`. I'll change it

divega commented on src/Microsoft.AspNet.Identity/LoggerExtensions.cs:
```diff
@@ -0,0 +1,74 @@
+
+namespace Microsoft.AspNet.Identity
+{
+    public static class LoggerExtensions
```
@suhasj When you go back to instance methods on UserManager, RoleManager and SignInManager, take a look at getting rid of this level of helpers. Seem largely unnecessary. You can do the check on the other ones and I don't think it is that much duplication.

AndriySvyryd commented on src/Microsoft.AspNet.Identity/LoggerExtensions.cs:
```diff
@@ -0,0 +1,74 @@
+            {
+                if (formatter != null)
+                {
+                    var message = await formatter(state, exception);
```
Can't the formatter be sync?

AndriySvyryd commented on src/Microsoft.AspNet.Identity/ResultExtensions.cs:
```diff
@@ -0,0 +1,102 @@
+        {
+            var logLevel = result ? LogLevel.Information : LogLevel.Warning;
+
+            await userManager.Logger.WriteAsync(logLevel, 0, methodName, null, async (callingMethod, exception) =>
```
Assuming that waiting for the write operation to finish isn't required to proceed with execution then it's not necessary to await on it.

suhasj replied:
so we call the method and continue. There won't be scenario when the execution fails because the calling thread is no longer executing right ?

AndriySvyryd replied:
If the call is not awaited on it will have nothing to do with the calling thread.
Also if an exception is thrown during execution of the called method it will not be observed unless there's a handler registered on TaskScheduler.UnobservedTaskException

suhasj commented:
@divega @HaoK 

Updated with review comments and this looks pretty good to me. I have corrected some doc comments as well

One possible stopper for checkin might be, right now the logging is done as 
`var baseMessage = Resources.Format(...)
Logger.WriteX(Resource.Format(baseMessage, args...)`

let me know if you want it as 
`Logger.WriteX(Resource.Format(Resources.Format(..,..), args...)`. I choose the first one just so it is more readable.

HaoK commented on src/Microsoft.AspNet.Identity/UserManager.cs:
```diff
@@ -1915,10 +1919,26 @@ public virtual Task<IList<TUser>> GetUsersInRoleAsync(string roleName)
+            }
             return result;
         }
 
```
I don't really like how we have two LogResult methods here, one for result, and one for bool.  Seems pretty clunky, this might be a sideeffect of removing the virtual Log methods from the various result objects, remind me why we removed the result virtual Log methods again?

suhasj replied:
The virtual Log method on the IdentityResult was not called by the bool overload anyway so they were present in the last checkin as well. The decision came because we constructed half message in UserManager/RoleManager and rest in the IdentityResult. This make convenient where do the whole construction of the message in one place after checking whether log level is enabled.

divega replied:
@HaoK Why would the result object need to know how to log? I prefer that it just knows how to return relevant data to construct the log message.

HaoK replied:
@divega Sure, I just mean the Result itself should have a say in how to format its log message, whether it does the final log, or generates a string for someone else to log doesn't matter, but the result object should have a say is my main point

divega replied:
Agreed, I just suggested that the logic to decide the status string (e.g. if (result.IsLockedOut) status = "Lockedout", etc) probably belongs in one of the result types. Maybe there are larger chunks that can be moved there.

divega commented on src/Microsoft.AspNet.Identity/UserManager.cs:
```diff
@@ -1929,17 +1949,23 @@ public virtual Task<IList<TUser>> GetUsersInRoleAsync(string roleName)
-                Logger.WriteInformation(string.Format("{0} : {1}", baseMessage, result.ToString()));
-            }
-            else
+            var logLevel = result ? LogLevel.Information : LogLevel.Warning;
```
This pattern looks a bit better than the one we have in the other methods, i.e. decide the log level, put it in a variable and checking if it is enabled instead of having nested ifs.

suhasj replied:
Sure. But we still need an additional check inside to call `LogInformation` or `LogWarning` appropriately based on the result but yes this reduces one line for sure

divega commented on src/Microsoft.AspNet.Identity/UserManager.cs:
```diff
@@ -1929,17 +1949,23 @@ public virtual Task<IList<TUser>> GetUsersInRoleAsync(string roleName)
-                Logger.WriteWarning(string.Format("{0} : {1}", baseMessage, result.ToString()));
+                var baseMessage = Resources.FormatLoggingResultMessage(methodName, await GetUserIdAsync(user));
+
+                if(logLevel == LogLevel.Warning)
```
Is there a version of `Write` we can use from logging that would just take `logLevel` as a parameter?

suhasj replied:
you mean like `Write(LogLevel level, string message)` ? no there isn't. The direct method off course takes eventid, state, exception and func

divega replied:
And accepts those to be null?

suhasj replied:
sure Logger.Write(LogLevel.Information, eventId, message, null, null) should work the same but yes not easily super concise. On the flip side, if we make the eventId configurable, then users can set specific eventId for Identity messages something not possible right now. Console logger doesn't much care about the eventId but other loggers might

divega replied:
We can chat about allowing the eventId configurable. The main reason I am asking right now bout this though is because you can replace the whole if with a single line of code.

divega commented on src/Microsoft.AspNet.Identity/SignInManager.cs:
```diff
@@ -415,15 +418,43 @@ public virtual AuthenticationProperties ConfigureExternalAuthenticationPropertie
+            {
+                var status = "";
+
+                if (result.IsLockedOut)
```
Consider moving this to a method in result that returns the right status string.

suhasj replied:
sure

divega commented on src/Microsoft.AspNet.Identity/RoleManager.cs:
```diff
@@ -387,14 +391,29 @@ private IRoleClaimStore<TRole> GetClaimStore()
         /// <param name="methodName"></param>
         /// <returns></returns>
-        protected async Task<IdentityResult> LogResultAsync(IdentityResult result,
+        protected virtual async Task<IdentityResult> LogResultAsync(IdentityResult result,
```
Can you refactor this method to also use something like `var logLevel = result.Succeeded ? LogLevel.Information : LogLevel.Warning` etc. as you did in a different version of the method?

suhasj replied:
sure I am doing it in all places

suhasj commented:
Updated with result classes to have a two string implementation.

One question pending is that the `LogResultAsync` (methods)[https://github.com/aspnet/Identity/pull/354/files#diff-bfbbcc7825aec8772885e70b17f60a84L409] in SignInManager  seems to have the same logic but operate on `bool` and `SignInResult` objects. They can be merged into a single method but the signature would `LogResultAsync(object result, ..)` which I feel is not recommended. Pointers on how to make this better.

Apart from that, added did some refactoring to static properties as suggested by @divega

HaoK commented on src/Microsoft.AspNet.Identity/SignInResult.cs:
```diff
@@ -40,74 +32,62 @@ public class SignInResult
-        {
-            get { return _failed; }
-        }
+        public static SignInResult Failed => new SignInResult();
```
This change now makes it so we are returning new instances every time the static property is called no?

divega replied:
Good point. My fault. I suggested @suhasj to look into something like this just to make this look simpler.

suhasj replied:
I had the same question but this change was suggested by Resharper as well. I can revert this back

divega replied:
We can just bring back the field and do `public static SignInResult Failed => _failed`

suhasj replied:
sure I was making that change :)

HaoK commented on src/Microsoft.AspNet.Identity/SignInResult.cs:
```diff
@@ -40,74 +32,62 @@ public class SignInResult
-        /// <param name="message"></param>
-        public virtual void Log(ILogger logger, string message)
+        /// <returns></returns>
+        public override string ToString()
```
You can make this read nicer with something like this
```
return (IsLockedOut ? "LockedOut" :
           IsNotAllowed ? "NotAllowed" :
           ""

HaoK replied:
Also the default should be Unknown as opposed to empty string.  Also shouldn't these go in a resource for loc?

suhasj replied:
Me and @divega had this chat and the consensus is that this is a behavior similar to enum so in that scenario it makes sense to hard code it.

divega replied:
I actually wasn't sure about it when we finished the conversation, so I think it is more of a question: These currently look like the implementation of `ToString()` for an enum, e.g. the strings represent members of the type in PascalCase. Should we return a descriptive name instead? If these are members of the type I don't think they should be localized. If they are descriptive strings then they should be. If they are not localized then I don't know if there is a good reason to still stick them in resources. @HaoK what do you think?

HaoK replied:
I don't feel super strongly about this one, we can go with them being more like an Enum and they should just match the boolean properties exactly then, since that should be easier

suhasj commented:
Updated

divega commented on src/Microsoft.AspNet.Identity/SignInResult.cs:
```diff
@@ -40,74 +38,43 @@ public class SignInResult
 
         /// <summary>
-        ///     Log result based on properties
+        ///     Returns string implemenation of the result. Provides string value for the assigned property
```
Implementation or representation?

divega replied:
The second sentence is redundant.

suhasj replied:
sure

divega commented on src/Microsoft.AspNet.Identity/SignInResult.cs:
```diff
@@ -40,74 +38,43 @@ public class SignInResult
+                ? "Lockedout"
+                : IsNotAllowed
+                    ? "NotAllowed"
+                    : RequiresTwoFactor ? "RequiresTwoFactor" : Succeeded ? "Succeeded" : "Failed";
```
Nit: I like the indentation and new lines to be the way it appeared in @HaoK's suggestion.

divega commented on src/Microsoft.AspNet.Identity/SignInManager.cs:
```diff
@@ -415,15 +418,20 @@ public virtual AuthenticationProperties ConfigureExternalAuthenticationPropertie
         /// <returns></returns>
-        protected async virtual Task<SignInResult> LogResultAsync(SignInResult status, TUser user, [System.Runtime.CompilerServices.CallerMemberName] string methodName = "")
+        protected virtual async Task<SignInResult> LogResultAsync(SignInResult result, TUser user,
+        [System.Runtime.CompilerServices.CallerMemberName] string methodName = "")
```
Nit: Add a using with the namespace of the attribute everywhere and make this shorter.

divega commented on src/Microsoft.AspNet.Identity/SignInManager.cs:
```diff
@@ -406,8 +405,12 @@ public virtual AuthenticationProperties ConfigureExternalAuthenticationPropertie
         {
-            Logger.WriteInformation(Resources.FormatLoggingSigninResult(Resources.FormatLoggingResultMessage(methodName,
-                await UserManager.GetUserIdAsync(user)), result));
+            // Check if log level is enabled before creating the message.
```
In UserManager we treat false as warning, e.g. `var logLevel = result ? LogLevel.Information : LogLevel.Warning`. Is it intentional that we treat true and false the same here?

suhasj replied:
I aligned the experience with the logging guidelines given at https://github.com/aspnet/Logging/wiki/Guidelines which says `Warning` is for unexpected flow in the application. In SignInManager, the result is bool for operations that are mostly read-only and most of the time it is an expected value. You might argue that locking out user might be an unexpected error but it is a known result. Hence I choose to log all flows of SignInManager. In contrast, adding an existing role to the database or creating a user with existing username is an error in the happy flow hence I chose it as Warning

I am open to this being changed since this is subjective. From a uniformity perspective yes, we can the same in SignInManager as well.

divega replied:
The guidelines are good information. I guess we need to decide when we apply them to Identity whether it is more helpful to log something like a user locked out as a warning. @HaoK what is your take?

HaoK replied:
So I would say, the two 'success' results for SignIn manager are Success and TwoFactorRequired, all of the other results should be considered failures and logged like user manager failures.

I would try True/False logging in the same way as UserManager

HaoK replied:
This sort of goes back to my point about the result object knowing how to render its log message, as today you'd have to put a switch inside SignInManager's Log to do this, while I think this decision should be taken care of by the result object (whether its currently a success or a failure).  That was why I origionally suggested for the Log to happen inside the result, while the managers method was mostly just to format the outer string and pass it into XXXResult.Log(outerformat), and the result call the appropriate Log method with warning/information since only the result knows truly whether its a failure or success

HaoK replied:
This is even more clear in the derived result case, when they extend SignInResult with an additional failure, they'd have to override both SignInResult.ToString and SignInManager.LogResult today, instead of just overriding SignInResult.ToString to add logic for any new status codes, and then deferring to the parent behavior

HaoK replied:
So I think we should add a test for each managers demonstrating a derived result class which adds a new status and updates the logging messages to handle them, that should make it easier to review what the extensibility story actually looks like, @suhasj can you add these tests as part of this PR?

HaoK commented on src/Microsoft.AspNet.Identity/SignInResult.cs:
```diff
@@ -40,74 +38,42 @@ public class SignInResult
-            {
-                logger.WriteInformation(Resources.FormatLoggingSigninResult(message, "Failed"));
-            }
+            return IsLockedOut ? "Lockedout" : 
```
nit: I'm not sure if this is the PR messed up or not, but can we make the indentation for all of the checks on the following line on the same indent level with IsLockedOut?

suhasj replied:
ah format document messed it up

HaoK commented:
Replaced by https://github.com/aspnet/Identity/pull/410 i  think
